File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
         ListNode *s=head,*f=head;
        while(f&&f->next){
            f=f->next->next;
            s=s->next;
            if(f==s){
                s=head;
                while(s!=f){
                    f=f->next;
                    s=s->next;
                }
                return s;
            }
        }
        
        return nullptr;
    }
};

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getFirstNode(ListNode head,ListNode slow){
        ListNode first=head;
        while(slow!=first){
            slow=slow.next;
            first=first.next;
        }
        return first;
    } 
    public ListNode detectCycle(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;
        while(fast!=null && fast.next!=null){
            
            fast=fast.next.next;
            slow=slow.next;
            if(slow==fast){
                return getFirstNode(head,slow);
            }
        }
        return null;
    }
}